From 726deff8e37229711f84d1694b3a2ade8f500723 Mon Sep 17 00:00:00 2001
From: Robert Nelson <robertcnelson@gmail.com>
Date: Thu, 17 Sep 2015 15:44:24 -0500
Subject: [PATCH 34/97] apply backport patch
 collateral-evolutions/network/0018-pv-trace-fixes/drivers_net_wireless_iwlwifi_iwl-debug.patch

---
 backports/drivers/net/wireless/iwlwifi/iwl-debug.c | 28 ++++++++++++++++------
 1 file changed, 21 insertions(+), 7 deletions(-)

diff --git a/backports/drivers/net/wireless/iwlwifi/iwl-debug.c b/backports/drivers/net/wireless/iwlwifi/iwl-debug.c
index 09feff4..c201d09 100644
--- a/backports/drivers/net/wireless/iwlwifi/iwl-debug.c
+++ b/backports/drivers/net/wireless/iwlwifi/iwl-debug.c
@@ -74,13 +74,16 @@ void __iwl_ ##fn(struct device *dev, const char *fmt, ...)	\
 	struct va_format vaf = {				\
 		.fmt = fmt,					\
 	};							\
-	va_list args;						\
+	va_list args1, args2;					\
 								\
-	va_start(args, fmt);					\
-	vaf.va = &args;						\
+	va_start(args1, fmt);					\
+	va_copy(args2, args1);					\
+	vaf.va = &args2;					\
 	dev_ ##fn(dev, "%pV", &vaf);				\
+	va_end(args2);						\
+	vaf.va = &args1;					\
 	trace_iwlwifi_ ##fn(&vaf);				\
-	va_end(args);						\
+	va_end(args1);						\
 }
 
 __iwl_fn(warn)
@@ -99,13 +102,18 @@ void __iwl_err(struct device *dev, bool rfkill_prefix, bool trace_only,
 	va_list args;
 
 	va_start(args, fmt);
-	vaf.va = &args;
 	if (!trace_only) {
+		va_list args2;
+
+		va_copy(args2, args);
+		vaf.va = &args2;
 		if (rfkill_prefix)
 			dev_err(dev, "(RFKILL) %pV", &vaf);
 		else
 			dev_err(dev, "%pV", &vaf);
+		va_end(args2);
 	}
+	vaf.va = &args;
 	trace_iwlwifi_err(&vaf);
 	va_end(args);
 }
@@ -122,13 +130,19 @@ void __iwl_dbg(struct device *dev,
 	va_list args;
 
 	va_start(args, fmt);
-	vaf.va = &args;
 #ifdef CONFIG_IWLWIFI_DEBUG
 	if (iwl_have_debug_level(level) &&
-	    (!limit || net_ratelimit()))
+	    (!limit || net_ratelimit())) {
+		va_list args2;
+
+		va_copy(args2, args);
+		vaf.va = &args2;
 		dev_printk(KERN_DEBUG, dev, "%c %s %pV",
 			   in_interrupt() ? 'I' : 'U', function, &vaf);
+		va_end(args2);
+	}
 #endif
+	vaf.va = &args;
 	trace_iwlwifi_dbg(level, in_interrupt(), function, &vaf);
 	va_end(args);
 }
-- 
2.5.1

